### ISSUE TITLE
Allow overriding of `classmethod`s with `staticmethod`s, and vice-versa

### ISSUE THREAD

**Issue by TTsangSC** [astral-sh/ty#4341]
### Summary

A base class providing an instance (resp. class) method should serve as an interface guaranteeing that the method is accessible on the instance (resp. class), and as such should allow subclasses to override them with e.g. class and static methods with a compatible signature when accessed on the aforementioned targets. This is how e.g. `mypy` handles substitution, and also what `ty` does in the case of overriding an instance method with a `classmethod`.

However, it seems that `staticmethod`s are handled differently (https://play.ty.dev/2762b5a2-b777-47e6-b705-20853dc7b441):
```python
from typing import Any, override


class Base:
    def instance_method(self) -> None:
        ...

    @classmethod
    def class_method(cls) -> None:
        ...


class OverrideWithClassMethod(Base):
    @override
    @classmethod
    def instance_method(cls) -> None:  # Okay
        ...


class OverrideWithStaticMethods(Base):
    @override
    @staticmethod
    def instance_method() -> None:  # "invalid-method-override"
        ...

    @override
    @staticmethod
    def class_method() -> None:  # "invalid-method-override"
        ...
```

Is this intended behavior or is it a bug? If a `staticmethod` cannot override an instance method (because of e.g. instance-bound side effects), shouldn't the override also be invalid for `classmethod`?

Cheers.

### Version

Playground (11c76bf48)

**Comment by carljm** [comment]
Thanks for the report!

I think not allowing a static method to override a class method is a bug; I don't know of any reasonable external way to observe that difference.

Instance methods are trickier, because Python (and ty) allow accessing them off class objects as unbound methods. Consider:

```py
class Base:
    def method(self) -> int:
        return 1

class Child(Base):
    @staticmethod
    def method() -> int:
        return 2

def call_unbound(cls: type[Base]) -> int:
    return cls.method(cls())

print(Child().method())     # 2: instance access works
print(call_unbound(Base))   # 1
print(call_unbound(Child))  # TypeError: Child.method() takes 0 positional arguments
                           # but 1 was given
```

But such calls to unbound methods are often already unsound anyway -- replace `cls()` with `Base()` in the body of `call_unbound` (which the type checker allows) and `call_unbound(Child)` would be unsound at runtime anyway, even if `Child` overrode with an instance method. So it's not clear to me how strong the rationale is for disallowing this, given that it's safe at the instance-type level.

It seems like currently our [documented intention](https://github.com/astral-sh/ruff/blob/4637a19fa01cca6b90e20904f2ed4d943a32cbe5/crates/ty_python_semantic/resources/mdtest/liskov.md?plain=1#L2229) is that the classmethod vs instancemethod case should be an error, but isn't yet.

Pyrefly, mypy, and zuban are OK with all overrides in the OP example; pyright errors on all of them.

I think I slightly lean towards the pyrefly/mypy/zuban approach here -- the only case in which these overrides are unsafe is calls to unbound methods accessed on a class, and those are inherently unsound to begin with. It's practically useful to be able to take advantage of the instance-level compatibility. So I would reverse the linked TODO and allow all of these.

@AlexWaygood what do you think?

**Comment by sharkdp** [comment]
After speaking with @AlexWaygood today (and seeing the ecosystem impact on a PR of mine), I think we should definitely allow the staticmethod vs classmethod case. This seems very common in the ecosystem.

Overriding instance methods with staticmethod/classmethod (or vice-versa) seems more questionable (as explained by Carl), and the use case is also less clear to me.

So I suggest that we implement the former behavior, and wait for more user feedback before deciding whether we want to implement the latter as well.

I'll repurpose this ticket for the former case. I've opened https://github.com/astral-sh/ty/issues/4504 to track the instance-method case separately.
